crates/ide-assists/src/handlers/convert_closure_to_fn.rs

```diff
@@ -0,0 +1,1258 @@
+use either::Either;
+use hir::{CaptureKind, ClosureCapture, FileRangeWrapper, HirDisplay};
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    base_db::SourceDatabase,
+    defs::Definition,
+    search::FileReferenceNode,
+    source_change::SourceChangeBuilder,
+    FxHashSet,
+};
+use std::fmt::Write;
+use syntax::{
+    algo::{skip_trivia_token, skip_whitespace_token},
+    ast::{
+        self,
+        edit::{AstNodeEdit, IndentLevel},
+        make, HasArgList, HasGenericParams, HasName,
+    },
+    hacks::parse_expr_from_str,
+    ted, AstNode, Direction, SyntaxKind, SyntaxNode, TextSize, ToSmolStr, T,
+};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: convert_closure_to_fn
+//
+// This converts a closure to a freestanding function, changing all captures to parameters.
+//
+// ```
+// # //- minicore: copy
+// # struct String;
+// # impl String {
+// #     fn new() -> Self {}
+// #     fn push_str(&mut self, s: &str) {}
+// # }
+// fn main() {
+//     let mut s = String::new();
+//     let closure = |$0a| s.push_str(a);
+//     closure("abc");
+// }
+// ```
+// ->
+// ```
+// # struct String;
+// # impl String {
+// #     fn new() -> Self {}
+// #     fn push_str(&mut self, s: &str) {}
+// # }
+// fn main() {
+//     let mut s = String::new();
+//     fn closure(a: &str, s: &mut String) {
+//         s.push_str(a)
+//     }
+//     closure("abc", &mut s);
+// }
+// ```
+pub(crate) fn convert_closure_to_fn(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    // The LetStmt needs to be first because it is applicable more widely.
+    let (text_range, closure_name, closure) = ctx
+        .find_node_at_offset::<ast::LetStmt>()
+        .and_then(|closure_decl| {
+            let closure_name = closure_decl.pat().and_then(|pat| match pat {
+                ast::Pat::IdentPat(pat) => pat.name().map(|name| (closure_decl.clone(), pat, name)),
+                _ => None,
+            });
+            Some((
+                closure_decl.syntax().text_range(),
+                closure_name,
+                match closure_decl.initializer()? {
+                    ast::Expr::ClosureExpr(it) => it,
+                    _ => return None,
+                },
+            ))
+        })
+        .or_else(|| {
+            ctx.find_node_at_offset::<ast::ClosureExpr>()
+                .map(|closure| (closure.syntax().text_range(), None, closure))
+        })?;
```

Comment:
We should only make this assist applicable on the closure signature imo (that is at or between the `|` bars), otherwise it'll clutter the assist dialog while within a closure, so `ctx.find_node_at_offset::<ast::ClosureExpr>()` + a filter for the cursor offset + check if the parent is a let statement